src/ai-jsx/util.ts

```diff
@@ -0,0 +1,367 @@
+
+export const AUTOBLOCKS_TRACKER_ID_PROP_NAME = 'autoblocks-tracker-id';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AnyComponent = Component<any>;
+
+/**
+ * Used to wrap a dynamic runtime variable so that the actual value is
+ * replaced in the template with {{ name }}.
+ */
+export function AutoblocksPlaceholder(props: AutoblocksPlaceholderProps) {
+  return props.children;
+}
+
+export function makeComponentName(f: AnyComponent): string {
+  return `<${f.name}>`;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function bindAsyncGenerator<T = unknown, TReturn = any, TNext = unknown>(
+  generator: AsyncGenerator<T, TReturn, TNext>,
+): AsyncGenerator<T, TReturn, TNext> {
+  const result = {
+    next: AsyncLocalStorage.bind(generator.next.bind(generator)),
+    return: AsyncLocalStorage.bind(generator.return.bind(generator)),
+    throw: AsyncLocalStorage.bind(generator.throw.bind(generator)),
+
+    [Symbol.asyncIterator]() {
+      return result;
+    },
+  };
+
+  return result;
+}
+
+/**
+ * Long story
+ *
+ * https://github.com/fixie-ai/ai-jsx/blob/4c67d845f48585dc3f26e90a9a656471f40c82ed/packages/ai-jsx/src/core/opentelemetry.ts#L91-L95
+ *
+ * This will break if AI.JSX changes the symbol name. 🤡
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function findMemoizedId(o: any): string | undefined {
+  try {
+    const symbols = Object.getOwnPropertySymbols(o);
+    const memoizedIdSymbol = symbols.find(
+      (s) => s.toString() === 'Symbol(memoizedId)',
+    );
+
+    if (memoizedIdSymbol) {
+      return `${o[memoizedIdSymbol]}`;
+    }
+    return undefined;
+  } catch {
+    return undefined;
+  }
+}
+
+function parseTrackerIdFromProps(props: unknown): string | undefined {
+  try {
+    const trackerId = (props as Record<string, unknown>)[
+      AUTOBLOCKS_TRACKER_ID_PROP_NAME
+    ];
+    if (typeof trackerId === 'string') {
+      return trackerId;
+    }
+    return undefined;
+  } catch {
+    return undefined;
+  }
+}
+
+function makeTemplateString(span: ABSpan | string): string {
+  if (typeof span === 'string') {
+    return span;
+  } else if (span.name === makeComponentName(AutoblocksPlaceholder)) {
+    const props = span.props as AutoblocksPlaceholderProps;
+    return `{{ ${props.name} }}`;
+  } else if (span.name === makeComponentName(OpenAIChatModel)) {
+    // There is a nested completion not wrapped in a placeholder
+    const name = parseTrackerIdFromProps(span.props) || 'completion';
+    return `{{ ${name} }}`;
+  } else {
+    return span.children.map(makeTemplateString).filter(Boolean).join('');
+  }
+}
+
+function makeMessageString(span: ABSpan | string): string {
+  if (typeof span === 'string') {
+    return span;
+  } else {
+    return span.children.map(makeMessageString).filter(Boolean).join('');
+  }
+}
+
+function countMessageTokens(span: ABSpan | string): number {
```

Comment:
i think the only tokens we can reliably count are the completion tokens, since they're streamed back one token at a time it's just the count of strings under an `<AssistantMessage />` component. for the prompt tokens (and thus total tokens) we'd have to use a 3rd party library, and the suggested one (tiktoken) is in python

Reply:
There is a js version, but i'm weary about adding it into the sdk because of the extra dependencies - https://www.npmjs.com/package/@dqbd/tiktoken

I think we'll likely want to add that calculation into the ingestion pipeline instead.

Reply:
yeah makes sense